Added pg@6 patching

jackhorton opened the pull request.

msftclas commented:
@jackhorton,
Thanks for your contribution as a Microsoft full-time employee or intern. You do not need to sign a CLA.
_Thanks,
Microsoft Pull Request Bot_

MSLaguana commented:
Looks like CI is failing to find tslint

jackhorton commented:
Yeah, something involving `npm link` was messing with it such that src/diagnostic-channel/node_modules/.bin didn't have a tslint executable. I changed the npm link stuff originally because the gulpfile used to link diagnostic channel to _itself_, which Jimmy and I agreed was incorrect. However, basically any other configuration of calling npm link, even if it worked on a clean build on my computer, seems to fail on Linux/CI. I've taken out all calls to link for now to make it pass, which should be fine since we weren't actually using it correctly before anyways.

The only reason I found the circular call to npm link was because it caused circular symlinks when trying to run the project in Docker.

MSLaguana commented on src/diagnostic-channel-publishers/src/postgres.pub.ts:
```diff
@@ -0,0 +1,128 @@
+                // emulate weird internal behavior in pg@6
+                // on success, the callback is called *before* query events are emitted
+                // on failure, the callback is called *instead of* the query emitting events
+                // with no events, that means no promises (since the promise is resolved/rejected in a )
```
Looks like you missed a a bit of the comment here.

MSLaguana commented on src/diagnostic-channel-publishers/src/postgres.pub.ts:
```diff
@@ -0,0 +1,128 @@
+                values = patchCallback(values);
+            }
+        } else {
+            if (typeof config.name === "string") {
```
What is the original behavior of `pgClient.query()` with no arguments? This check here will cause an exception if `config` is null/undefined.

jackhorton replied:
It immediately looks at `config.submit` before anything else, so null/undefined is already an issue:

https://github.com/brianc/node-postgres/blob/v6.4.1/lib/client.js#L355

MSLaguana commented:
Did you find any scenarios where the default `pg` module did not preserve zone prior to these changes? E.g. starting multiple queries in quick succession, possibly using a pool? It'd be good to see a test for a scenario that would fail without the patch.

mike-kaufman commented on src/diagnostic-channel-publishers/src/postgres.pub.ts:
```diff
@@ -0,0 +1,128 @@
+        let queryResult;
+
+        function patchCallback(cb?: PostgresCallback): PostgresCallback {
+            return channel.bindToContext(function trackingCallback(err: Error, res: IPostgresResult): any {
```
I'm wondering if you need some logic here to detect already wrapped callbacks, and then unwrap/re-wrap them if necessary.   e.g., this probably won't work exactly as you intended...
```javascript
const callback = () => {...};
config = { callback };
postgress.query(config);  // this will wrap up config.callback
postgress.query(config); // this will wrap up config.callback again!
```

jackhorton replied:
Good call, I can add that.

jackhorton replied:
So, I don't actually need to worry about them being double-wrapped, since I'm creating a new function object and even if the user passes in the same function twice, I would just create a new object again. With that being said, that seems wasteful, so I am adding a WeakMap<userFunc, patchedFunc> which is slightly more overhead for us but significantly less overhead in case anyone is using this pattern for any reason (id imagine libraries wrapping pg very likely could use this).

mike-kaufman commented on src/diagnostic-channel-publishers/src/postgres.pub.ts:
```diff
@@ -0,0 +1,128 @@
+        // EventEmitter returned (well, sometimes -- see above), its safe to insert a callback if none was given
+        if (typeof config === "string") {
+            if (values instanceof Array) {
+                data.query.preparable = {
```
Consider making a copy of any objects before mutating them.   Will prevent any unintended side-affects when users are doing "something else" with these objects.

jackhorton replied:
I may be having a reference-related misunderstanding here, but if I'm creating a local `data` variable on each call to `client.query()`, I can't possibly be mutating data objects that anyone else could have reference to, no?

mike-kaufman replied:
so, this you're correct:

```
function f(config) {
   const obj = {};  
   obj.x = 'y';
}
```

this though would modify the instance the user passed in:

```
function f(config) {
   const obj = config;
   obj.x = 'y';
}
```

lmk if we're talking past each other...

mike-kaufman commented on src/diagnostic-channel-publishers/src/postgres.pub.ts:
```diff
@@ -0,0 +1,128 @@
+            }
+        }
+
+        queryResult = originalClientQuery.call(this, config, values, callback);
```
should probably account for any additional args passed along here (ie..e, if I call with 4 or 5 args, those should get propagated through.  E.g., `queryResult = originalClientQuery.apply(this, arguments);`  Also, hard to tell in the CR with the comments, but I'm assuming `this` &  `arguments` are still correct for the context of when original `query` was called.

jackhorton replied:
The original module never uses arguments internally, but I also assumed that reassigning config/values/callback wouldn't update the arguments array, but I just checked and it does. I can fix this.

jackhorton replied:
Actually, I was wrong. I am not sure why my quick test worked but switching to apply(this, arguments) here breaks things. I could switch every argument reference to an `arguments[x]` for some `x`, but I am very sure that pg, at least in version 6, doesn't use anything more than the three it was provided.

mike-kaufman commented on src/diagnostic-channel-publishers/src/pg.pub.ts:
```diff
@@ -0,0 +1,164 @@
+    // this keeps track of user-supplied callback => patched callbacks so we don't keep
+    // re-patching the same callback over and over for library functions
+    // IMPORTANT: use a WeakMap so that we don't keep references to temporary/otherwise unreferenced functions
+    const patchedCallbacks = new WeakMap<PostgresCallback, PostgresCallback>();
```
so, is WeakMap going to work on down-level node versions?  Better approach may be to tag the function with the original, e.g., something like this:

```
function stamp(cb, wrapper) {
   cb.__diagnosticChannelProxy = wrapper;
   wrapper.__diagnosticChannelOriginal = cb;
}
```
 
Then you can easily detect if any given function has been proxied, and you can easily traverse between proxy & proxied function.  And it works across node versions with no concerns about memory leaks. 

Also, should probably be consistent w/ this approach in all the functions we proxy.

jackhorton replied:
Ah, I suppose supporting ES5 is important for node < 6, I didn't even think of that. I was using that originally but I thought the weakmap would be more clever.

mike-kaufman commented on src/diagnostic-channel-publishers/src/pg.pub.ts:
```diff
@@ -0,0 +1,164 @@
+                    patchedCallbacks.set(existingPatch, existingPatch);
+                }
+
+                return existingPatch;
```
don't think this will necessarily use the correct context here.  It will use the context that the original patch was created for, not the current context.

mike-kaufman commented on src/diagnostic-channel-publishers/src/pg.pub.ts:
```diff
@@ -0,0 +1,164 @@
+
+            const trackingCallback = channel.bindToContext(function(err: Error, res: IPostgresResult): any {
+                const end = process.hrtime(start);
+                data.result = res && { rowCount: res.rowCount, command: res.command };
```
We should be watching what values are in `res.command`.  e.g., we probably don't want to send commands like `alter user myuser set  password = "xxx"`...

jackhorton replied:
res.command is always the command "type" (at least according to [pg v7](https://node-postgres.com/api/result#-code-result-command-string-code-), I can double-check v6 but I don't remember this changing much), so in this case it would be just the string "ALTER"

mike-kaufman commented on src/diagnostic-channel-publishers/src/pg.pub.ts:
```diff
@@ -0,0 +1,164 @@
+                }
+            });
+
+            if (cachePatched) {
```
not following why this `if (cachPatched)` block is necessary here...

mike-kaufman commented on src/diagnostic-channel-publishers/src/pg.pub.ts:
```diff
@@ -0,0 +1,164 @@
+            return trackingCallback;
+        }
+
+        // this function takes too many variations of aguments.
```
sp: arguments

mike-kaufman commented on src/diagnostic-channel-publishers/src/pg.pub.ts:
```diff
@@ -0,0 +1,164 @@
+                }
+            }
+        } else {
+            if (typeof config.name === "string") {
```
if `config === undefined || null`, then this blows up.  Also, what if config `typeof config === 'number'`?  May be useful to look at what the postgres driver is doing to normalize args passed in here, and just copy that.  E.g., you could have a function like this:

```typescript
function normalizeQueryArgs(config, values, callback): Object {
  if ( !config || (typeof config !== 'string' && typeof config !== 'number) {
    // unsupported types, don't proxy these, just let pg driver error out however it errors out
    return config;
  }
  else if (typeof config === 'string') {
config.xx = ...
  }
  else {

  }  
  return config; // now all subsequent logic can just deal w/ a `config` object, which simplifies things.
  }
}
```

jackhorton replied:
So this is all based on that normalizeQueryArgs loosely, but I didn't want to do what that function does and create my own config object because I wanted to pass in the args as closely to the original call as possible (aside from possibly adding a callback to the end of the args list if need be).

mike-kaufman commented on src/diagnostic-channel-publishers/src/pg.pub.ts:
```diff
@@ -0,0 +1,164 @@
+            }
+        }
+
+        queryResult = originalClientQuery.call(this, config, values, callback);
```
this should always call `originalClient.apply(this, arguments)`, as we want to preserve whatever the user is doing, right or wrong.  What's happening here is, if you user  is calling this:

```
   pg.query(myConfig, myValues, myCallback, 1, 2, 3, 4, 5, 6, 7, 8); 
```

You're stripping off the last 8 arguments.  Now, it should matter, but in principle our proxies shouldn't have side affects like this. 

Also note that if call `(originalQuery.apply(this, arguments)`, you need to make sure you update the arguments array with whatever you've replaced.

jackhorton commented on src/diagnostic-channel-publishers/src/pg-pool.pub.ts:
```diff
@@ -0,0 +1,27 @@
+
+    originalPgPool.prototype.connect = function connect(callback?: Function): void {
+        callback = callback
+            ? channel.bindToContext(callback)
```
Do you think this unwrap-rewrap should just be part of the `bindToContext` API? I feel like this would be a concern nearly anywhere. For what its worth, it does currently work, at least when the context preserving function is Zone.wrap, but it might not if the context preserver has side effects.

mike-kaufman commented:
> Do you think this unwrap-rewrap should just be part of the bindToContext API? I feel like this would be a concern nearly anywhere. For what its worth, it does currently work, at least when the context preserving function is Zone.wrap, but it might not if the context preserver has side effects.

the wrap/unwrap/iswrapped logic is probably best determined by the context implementation.  The concern I have, including with zones, is something like the following (waving hands a bit): 

```
let someFunc = () = {
    console.log('my context is: ' + getCurrentContext());
};
...
//  context A
someFunc = Zone.wrap(someFunc);
...
// context B
someFunc = Zone.wrap(someFunc);    // someFunc is already wrapped to exec in A, so this doesn't have correct result.
```

jackhorton commented:
I see what you mean with the zone wrapping concerns, but that wouldn't happen in the case you pointed out in `Pool.prototype.connect` because the user-supplied callback will never be reassigned to the newly-bound callback, since that's all internal. Or am I misunderstanding?

I added a test to confirm that creating a .connect() handler in the root zone and then using it in two different zones properly makes the callback run in the zone that is currently calling connect, and I also added property-based checking to access the "original" user callback in the `config: { callback: fn }` scenario for .query(), as well as adding tests to confirm those work as intended (though, the tests are fragile enough that I'd appreciate if you took a look at them).

mike-kaufman commented on src/diagnostic-channel-publishers/src/pg.pub.ts:
```diff
@@ -0,0 +1,160 @@
+            }
+        } catch (e) {
+            // if our logic here throws, bail out and just let pg do its thing
+            return originalClientQuery.apply(this, arguments);
```
something to think about going forward - for cases like this is an error hook.  Clients of this library (e.g., app insights) should be hooking sending telemetry when we fall into blocks like this.

jackhorton replied:
Id guess that would be on a new channel?

```ts
export interface IPatchError {
  module: string,
  argConfig: string[] // typeof each arg
}
channel.publish("patch-error", { module: "pg", argConfig: Array.prototype.slice.call(arguments).map(arg => typeof arg) });
```

We wouldn't want to include actual args since those might contain sensitive info, and this would require a bit more work with objects to inspect them more.

mike-kaufman replied:
Per offline discussion, this should be fine.  Alternatively, you can make this either an event or just an interface that clients cal call e.g., `registerErrorHandler(handler: IErrorHandler)`.  Key thing to think though is is an error need to be on the same notification as a "channel" (where we're firing events that are conceptually different than "somethign didn't work").  I haven't though this.   Plz open a bug to track.

mike-kaufman commented on src/diagnostic-channel-publishers/src/pg.pub.ts:
```diff
@@ -0,0 +1,160 @@
+        arguments[0] = config;
+        arguments[1] = values;
+        arguments[2] = callback;
+        arguments.length = (arguments.length > 3) ? arguments.length : 3;
```
hmmm, this is little weird.  It may be mroe clear here to update line 38 like this:

`originalPg.Client.prototype.query = function query(config, values, callback, rest...) {`

and then  line 146 like this:

`queryResult = originalQuery.call(config, values, callback, rest...) {`

This is syntactic sugar that should get transpiled away & work on down-level runtimes.

jackhorton replied:
I can make another PR that adds this for all of the modules? This sort of arguments trickery is needed whenever we want to support calling with more than the named arguments _and_ fewer than the named arguments.

mike-kaufman replied:
Per offline discussion, this is fine.
